File: app/src/main/java/me/yxp/qfun/hook/api/OnTroopQuit.java
package me.yxp.qfun.hook.api;

import org.json.JSONObject;

import java.lang.reflect.Method;
import java.util.ArrayList;

import kotlin.collections.CollectionsKt;
import me.yxp.qfun.hook.base.ApiHookItem;
import me.yxp.qfun.utils.hook.HookUtils;
import me.yxp.qfun.utils.json.ProtoData;
import me.yxp.qfun.utils.qq.HostInfo;
import me.yxp.qfun.utils.qq.QQUtils;
import me.yxp.qfun.utils.reflect.ClassUtils;
import me.yxp.qfun.utils.reflect.MethodUtils;

public final class OnTroopQuit extends ApiHookItem {
    public static final ApiHookItem INSTANCE = new OnTroopQuit();

    @Override
    public void loadHook() throws Throwable {
        if (HostInfo.isTIM()) {
            Method handleMemberExit = MethodUtils.create(ClassUtils._TroopOnlinePushHandler())
                    .withMethodName("handleMemberExit")
                    .findOne();
            HookUtils.hookAlways(handleMemberExit, null, param -> handleQuit((String) param.args[0], (String) param.args[1]));
        }

        if (HostInfo.isQQ()) {
            Class<?> troopMemberRemovePushProcessor = ClassUtils
                    .load("com.tencent.qqnt.push.processor.TroopMemberRemovePushProcessor");
            Method quitGroup = MethodUtils.create(troopMemberRemovePushProcessor)
                    .withReturnType(void.class)
                    .withParamTypes(ArrayList.class)
                    .findOne();

            HookUtils.hookAlways(quitGroup, null, param -> {
                ArrayList<Byte> byteList = (ArrayList<Byte>) param.args[0];
                ProtoData data = new ProtoData();
                data.fromBytes(CollectionsKt.toByteArray(byteList));
                JSONObject json = data.toJSON();

                String troopUin = String.valueOf(json.getJSONObject("1").getLong("1"));
                String memberUid = json.getJSONObject("3").getJSONObject("2").getString("3");
                String memberUin = QQUtils.getUinFromUid(memberUid);
                handleQuit(troopUin, memberUin);

            });
        }
    }

    private void handleQuit(String troopUin, String memberUin) throws Throwable {
        for (Listener listener : mListenerSet) {
            if (listener instanceof TroopQuitListener) {
                ((TroopQuitListener) listener).onQuit(troopUin, memberUin);
            }
        }
    }

    public interface TroopQuitListener extends Listener {
        void onQuit(String troopUin, String memberUin) throws Throwable;
    }
}

File: app/src/main/java/me/yxp/qfun/hook/api/OnTroopQuit.kt
package me.yxp.qfun.hook.api

import com.tencent.mobileqq.troop.api.impl.TroopMemberInfoServiceImpl
import me.yxp.qfun.annotation.HookItemAnnotation
import me.yxp.qfun.hook.base.BaseApiHookItem
import me.yxp.qfun.hook.base.Listener
import me.yxp.qfun.utils.hook.hookAfter

@HookItemAnnotation("监听用户退群")
object OnTroopQuit : BaseApiHookItem<TroopQuitListener>() {
    override fun loadHook() {

        TroopMemberInfoServiceImpl::class.java
            .getDeclaredMethod(
                "deleteTroopMember",
                String::class.java,
                String::class.java,
                Boolean::class.javaPrimitiveType
            )
            .hookAfter(this) { param ->
                val troopUin = param.args[0] as String
                val memberUin = param.args[1] as String
                listenerSet
                    .filter {
                        verify(it)
                    }.forEach {
                        it.onQuit(troopUin, memberUin)
                    }
            }


    }
}

fun interface TroopQuitListener : Listener {
    fun onQuit(troopUin: String, memberUin: String)
}